### PR TITLE
Add type hints for arelle/ValidateXbrl.py and ancillary files

### PR DESCRIPTION
**Pull request by ggravlingen**
#### Reason for change
We add more type hints.

#### Description of change
This PR:
- Focuses on adding type hints for `arelle/ValidateXbrl.py`. There are some places where I might have used `Any`. In those cases, I wasn't sure and felt we'll probably be able to clarify the type down the line as we add more files to the library of typed files.
- Also adds type hints for ancillary files. I've focused on return types to somewhat reduce the complexity of the PR.

#### Steps to Test

**review**:
@Arelle/arelle

### CODE REVIEW

**Comment by aviary2-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: arelle/ValidateXbrl.py
Comment (ggravlingen) [inline arelle/ValidateXbrl.py:67]: If you (who have more experience with using the library) know what is included in these sets, please let me know!

**Comment by ggravlingen** [comment]
I'll look into the failing tests.

File: arelle/ValidateXbrl.py
Comment (selgamal) [reply arelle/ValidateXbrl.py:66]: domain members are model concepts --> set[ModelConcept]
extensionImportedUrls is string, modelDocument.uri --> set[str]

File: arelle/ValidateXbrl.py
Comment (ggravlingen) [reply arelle/ValidateXbrl.py:170]: I have given this some thought over the day, also considering that tests fail, and I agree. This is a bad idea that alters the behaviour of the code. I'll follow your advise.

File: arelle/ValidateXbrl.py
Comment (ggravlingen) [reply arelle/ValidateXbrl.py:934]: My understanding was that `dict_values` (which what's passed to the argument) is to be typed as `ValuesView`. I'll give `Iterable` a go and see if mypy accepts that.

File: arelle/ValidateXbrl.py
Comment (ggravlingen) [reply arelle/ValidateXbrl.py:66]: Thanks!

File: arelle/ValidateXbrl.py
Comment (ggravlingen) [reply arelle/ValidateXbrl.py:266]: I was actually trying to solve these typing errors:
```
arelle/ValidateXbrl.py:265: error: "error" of "ModelXbrl" does not return a value  [func-returns-value]
```

But you are right, those bugs can be fixed in a follow-up. I'll revert and add a type: ignore.

File: arelle/ModelXbrl.py
Comment (ggravlingen) [reply arelle/ModelXbrl.py:304]: It's kind of nice to let mypy know it's a dict. In the future, we should of course be more specific of what it contains.

File: arelle/ValidateXbrl.py
Comment (ggravlingen) [reply arelle/ValidateXbrl.py:833]: I was again trying to solve this:
```
arelle/ValidateXbrl.py:861: error: "error" of "ModelXbrl" does not return a value  [func-returns-value]
```

File: arelle/ValidateXbrl.py
Comment (ggravlingen) [reply arelle/ValidateXbrl.py:934]: No, you were of course right :)

**Comment by selgamal** [comment]
Is anyone working on Cntlr.py? 
I already done type hinting and tested  a part of that file in a local branch I can complete it and push it if no one is working on that file

**Comment by ggravlingen** [comment]
At least I'm not working on it and I think the Arelle team are busy with some other things, so I guess it's up for grabs!

File: arelle/ValidateXbrl.py
Comment (ggravlingen) [reply arelle/ValidateXbrl.py:833]: Ah, nice catch. I'll remove that.

File: arelle/ValidateXbrl.py
Comment (ggravlingen) [reply arelle/ValidateXbrl.py]: Sorry, I misunderstood you here.

**Comment by selgamal** [comment]
@stevenbronson-wk well its one way to spend the morning
But how do I test my work locally, I tried  `mypy arelle --namespace-packages`, `mypy -m arelle.Cntrl`, and `mypy -p arelle`, not getting the result I expected

**Comment by ggravlingen** [comment]
@selgamal please push your branch and open a new (draft) PR so we can keep the discussion here about the proposed changes here.

To answer your question, you need to add the file to pyproject.toml. Have a look at the recent PR for the updater.

**Comment by selgamal** [comment]
> @selgamal please push your branch and open a new (draft) PR so we can keep the discussion here about the proposed changes here.
> 
> To answer your question, you need to add the file to pyproject.toml. Have a look at the recent PR for the updater.

Got it! Thank you

File: arelle/ValidateXbrl.py
Comment (ggravlingen) [reply arelle/ValidateXbrl.py]: I'll fix this in a follow-up.

File: arelle/ValidateXbrl.py
Comment (ggravlingen) [reply arelle/ValidateXbrl.py:660]: This is what VS code suggests:
(variable) toLabel: Unbound | str | Any | None

File: arelle/ValidateXbrl.py
Comment (ggravlingen) [reply arelle/ValidateXbrl.py:665]: VS Code suggests: `(variable) toLabel: str` on my side. But it might be wrong. Are you spotting anything else?

File: arelle/ValidateXbrl.py
Comment (ggravlingen) [reply arelle/ValidateXbrl.py:1045]: Sorry, I messed up when reverting.

File: arelle/ValidateXbrl.py
Comment (ggravlingen) [reply arelle/ValidateXbrl.py]: I see this is still marked as unresolved. Did I miss to address anything here?

File: tests/integration_tests/validation/XBRL/test_xbrl_conformance_suite.py
Comment (ggravlingen) [reply tests/integration_tests/validation/XBRL/test_xbrl_conformance_suite.py:5]: Ooops, I had to remove `.zip` to get the test suite to run on my mac. Will revert.

Comment (ggravlingen) [reply tests/integration_tests/validation/XBRL/test_xbrl_conformance_suite.py]: I have a bad habit of running `git add .` :(

**Comment by ggravlingen** [comment]
Tests will fail here, still looking into why. Can't spot any obvious breaking changes.

File: arelle/XmlUtil.py
Comment (ggravlingen) [inline arelle/XmlUtil.py:380]:
```suggestion
def descendant(element, descendantNamespaceURI, descendantLocalNames, attrName=None, attrValue=None) -> Any:
```

File: arelle/ValidateXbrl.py
Comment (ggravlingen) [reply arelle/ValidateXbrl.py]: Ah, sweet. I was running the tests locally and still getting errors but it looks like they're passing here. Should probably clean this commit history up a bit before you merge this.

Comment (ggravlingen) [reply arelle/ValidateXbrl.py]: I ran a command similar to this the other day to clean up: `git rebase -i d668263c~` The commit hash, should that be the first in the series of commits in this PR?

**Comment by ggravlingen** [comment]
Trying to find my way around how to clean up the commit history here. I made a backup of the branch before I started in case I would mess up.

File: arelle/ValidateXbrl.py
Comment (sagesmith-wf) [inline arelle/ValidateXbrl.py:611]: This seems like arelle needs to handle gracefully. Probably produce a schema error or something appropriate? Here and other places. Reasoning is that arelle is being used to validate docs so it _should_ handle and report any error it finds with a filing back to the user

File: arelle/ModelXbrl.py
Comment (ggravlingen) [reply arelle/ModelXbrl.py:271]: This is intentional and this is my rationale: in this PR, I propose type hints for `ValidateXbrl`. `ModelXbrl` not in direct scope, so I'm not being super accurate when adding type hints here. When revisiting this file, however, I (or whoever works on it) should be more accurate.

File: arelle/PluginManager.py
Comment (ggravlingen) [reply arelle/PluginManager.py:25]: I'd say this is up to you guys and the what guidelines you want us to work under but I don't think I've seen those imports under `if TYPE_CHECKING` before.

File: arelle/ModelXbrl.py
Comment (austinmatherne-wk) [reply arelle/ModelXbrl.py:1232]: Looking at the actual usage of codes in `effectiveMessageCode()` it's a bug for this to be a list. It should be `str | tuple[str, ...]`

File: arelle/PluginManager.py
Comment (austinmatherne-wk) [reply arelle/PluginManager.py:25]: The import overhead seems minimal (most of these are going to be imported anyway, so it's typically a cheap hash lookup). I'd prefer to limit `TYPE_CHECKING` usage. Use it to avoid circular dependencies and if we actually discover that something introduces overhead.

File: arelle/ValidateXbrl.py
Comment (austinmatherne-wk) [reply arelle/ValidateXbrl.py:611]: Agreed. @ggravlingen you don't need to handle error logging here (kind of out of scope for a type hints pr), but I'd avoid adding this assertion. Replace it with an if statement to guard against a missing arcrole:
```pytohn
if arcrole is not None:
    self.genericArcArcroles.add(arcrole)
    if arcrole in (XbrlConst.elementLabel, XbrlConst.elementReference):
        resourceArcTos.append((toLabel, arcrole, arcElt))
```

The other None assertions seem fine (they'd cause errors anyway.)

Comment (austinmatherne-wk) [reply arelle/ValidateXbrl.py:611]: Thanks @stevenbronson-wk, I missed that comment.

I did a little investigation and the arcrole can't be None here (Arelle blows up earlier). Ultimately 2 would be best, but I'm okay with the assertion.

```
[AttributeError] Unrecoverable error: 'NoneType' object has no attribute 'startswith', wk-20220331.htm - ../IXDS wk-20220331.htm
Traceback (most recent call last):
  File "/Users/austinmatherne/git/arelle/arelle/ModelDocument.py", line 146, in load
    modelDocument = pluginMethod(modelXbrl, normalizedUri, filepath, isEntry=isEntry, namespace=namespace, **kwargs)
  File "/Users/austinmatherne/git/arelle/arelle/plugin/inlineXbrlDocumentSet.py", line 121, in inlineXbrlDocumentSetLoader
    ixdoc = load(modelXbrl, elt.text, referringElement=elt, isDiscovered=True)
  File "/Users/austinmatherne/git/arelle/arelle/ModelDocument.py", line 332, in load
    modelDocument.inlineXbrlDiscover(rootNode)
  File "/Users/austinmatherne/git/arelle/arelle/ModelDocument.py", line 1326, in inlineXbrlDiscover
    self.schemaLinkbaseRefsDiscover(inlineElement)
  File "/Users/austinmatherne/git/arelle/arelle/ModelDocument.py", line 1088, in schemaLinkbaseRefsDiscover
    self.schemaLinkbaseRefDiscover(element)
  File "/Users/austinmatherne/git/arelle/arelle/ModelDocument.py", line 1095, in schemaLinkbaseRefDiscover
    return self.discoverHref(element, urlRewritePluginClass="ModelDocument.InstanceSchemaRefRewriter")
  File "/Users/austinmatherne/git/arelle/arelle/ModelDocument.py", line 1215, in discoverHref
    doc = _newDoc(self.modelXbrl, url, isDiscovered=not nonDTS, base=self.baseForElement(element), referringElement=element)
  File "/Users/austinmatherne/git/arelle/arelle/ModelDocument.py", line 325, in load
    modelDocument.schemaDiscover(rootNode, isIncluded, isSupplemental, namespace)
  File "/Users/austinmatherne/git/arelle/arelle/ModelDocument.py", line 925, in schemaDiscover
    self.schemaDiscoverChildElements(rootElement, isSupplemental)
  File "/Users/austinmatherne/git/arelle/arelle/ModelDocument.py", line 981, in schemaDiscoverChildElements
    self.schemaDiscoverChildElements(modelObject, isSupplemental)
  File "/Users/austinmatherne/git/arelle/arelle/ModelDocument.py", line 981, in schemaDiscoverChildElements
    self.schemaDiscoverChildElements(modelObject, isSupplemental)
  File "/Users/austinmatherne/git/arelle/arelle/ModelDocument.py", line 973, in schemaDiscoverChildElements
    self.schemaLinkbaseRefDiscover(modelObject)
  File "/Users/austinmatherne/git/arelle/arelle/ModelDocument.py", line 1095, in schemaLinkbaseRefDiscover
    return self.discoverHref(element, urlRewritePluginClass="ModelDocument.InstanceSchemaRefRewriter")
  File "/Users/austinmatherne/git/arelle/arelle/ModelDocument.py", line 1215, in discoverHref
    doc = _newDoc(self.modelXbrl, url, isDiscovered=not nonDTS, base=self.baseForElement(element), referringElement=element)
  File "/Users/austinmatherne/git/arelle/arelle/ModelDocument.py", line 327, in load
    modelDocument.linkbaseDiscover(rootNode)
  File "/Users/austinmatherne/git/arelle/arelle/ModelDocument.py", line 1169, in linkbaseDiscover
    if XbrlConst.isDimensionArcrole(arcrole) and not dimensionArcFound:
  File "/Users/austinmatherne/git/arelle/arelle/XbrlConst.py", line 763, in isDimensionArcrole
    return arcrole.startswith("http://xbrl.org/int/dim/arcrole/")
AttributeError: 'NoneType' object has no attribute 'startswith'
```

File: .gitignore
Comment (ggravlingen) [reply .gitignore:41]: Guess you already have it in your local global gitignore?

File: arelle/ModelXbrl.py
Comment (ggravlingen) [reply arelle/ModelXbrl.py:275]: Addressed in another comment.

File: arelle/ValidateXbrl.py
Comment (ggravlingen) [reply arelle/ValidateXbrl.py:996]: Certainly. This is my preferred method of splitting long lines (and how Black would have formatted it).

File: arelle/ValidateXbrl.py
Comment (ggravlingen) [reply arelle/ValidateXbrl.py:611]: I take it I should leave this as is then?

File: arelle/ValidateXbrl.py
Comment (ggravlingen) [reply arelle/ValidateXbrl.py:177]: Good idea!

File: arelle/ValidateXbrlDimensions.py
Comment (ggravlingen) [reply arelle/ValidateXbrlDimensions.py:541]: Yes, you were right, this was not necessary.

**Review by austinmatherne-wk — APPROVED** [review]

**Pull request merged**
